sound/soc/sof/ipc4-pcm.c

```diff
@@ -462,7 +462,91 @@ static int sof_ipc4_pcm_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm
 	return 0;
 }
 
+static void sof_ipc4_build_time_info(struct snd_sof_dev *sdev, struct snd_sof_pcm_stream *spcm)
+{
+	struct sof_ipc4_copier *host_copier = NULL;
+	struct sof_ipc4_copier *dai_copier = NULL;
+	struct sof_ipc4_llp_reading_slot llp_slot;
+	struct sof_ipc4_timestamp_info *info;
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_dai *dai;
+	int i;
+
+	/* find host & dai to locate info in memory window */
+	for_each_dapm_widgets(spcm->list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+
+		if (!swidget)
+			continue;
+
+		if (WIDGET_IS_AIF(swidget->widget->id)) {
+			host_copier = swidget->private;
+		} else if (WIDGET_IS_DAI(swidget->widget->id)) {
+			dai = swidget->private;
+			dai_copier = dai->private;
+		}
+	}
+
+	info = spcm->private;
+	info->host_copier = host_copier;
+	info->dai_copier = dai_copier;
+	info->llp_offset = offsetof(struct sof_ipc4_fw_registers, llp_gpdma_reading_slots) +
+				    sdev->fw_info_box.offset;
+
+	/* find llp slot used by current dai */
+	for (i = 0; i < SOF_IPC4_MAX_LLP_GPDMA_READING_SLOTS; i++) {
+		sof_mailbox_read(sdev, info->llp_offset, &llp_slot, sizeof(llp_slot));
+		if (llp_slot.node_id == dai_copier->data.gtw_cfg.node_id)
+			break;
+	}
+
+	if (i == SOF_IPC4_MAX_LLP_GPDMA_READING_SLOTS) {
+		/* if no llp gpdma slot is used, check aggregated sdw slot */
+		info->llp_offset = offsetof(struct sof_ipc4_fw_registers, llp_sndw_reading_slots) +
+						sdev->fw_info_box.offset;
+		for (i = 0; i < SOF_IPC4_MAX_LLP_SNDW_READING_SLOTS; i++) {
+			sof_mailbox_read(sdev, info->llp_offset, &llp_slot, sizeof(llp_slot));
+			if (llp_slot.node_id == dai_copier->data.gtw_cfg.node_id)
+				break;
+		}
+
+		/* no llp support */
+		if (i == SOF_IPC4_MAX_LLP_SNDW_READING_SLOTS) {
```

Comment:
can it be in the `llp_evad_reading_slot`?

Reply:
Currently cSOF doesn't support it.  At first SOF_IPC4_MAX_LLP_SNDW_READING_SLOTS was defined as 16, and
SOF_IPC4_MAX_LLP_SNDW_READING_SLOTS - 1 is used for SNDW, and the 1 is used for EVAD which is also supported by SNDW. To support it by one shot.  It is better to check it.

Change SOF_IPC4_MAX_LLP_SNDW_READING_SLOTS back to 16, and also check EVAD slot.

Reply:
right, this is really hackish and unsafe, imho.
If we don't care if it is EVAD or not EVAD, then don't mention it and have the sndw slots to be 16 and that's it, don't mention EVAD.

Reply:
so you prefer still use 15 for SOF_IPC4_MAX_LLP_SNDW_READING_SLOTS, and don't care EVAD now ?